###### Question 1: How to read pandas doc? :

  Tip1: Using the pandas documentation is to get to right page and generally the easiest way is to google it rather than going directly to the pandas documents and then searching. 
    Just go to Google -> (type) pandas functionname ->hit enter. eg: type "pandas read_csv " hit enter.
Now we will see the results and we need to make sure to choose the newest version(right now it is 1.2.2). 
So will go head and click on the newest version. When the first page opens and when we see the name of the function 
it just says pandas.read_csv() and its telling us its a Top/Higher-Level function(The closer it is to human language, the higher-level the function is.) eg: High level functions: Car.Start() or Car.MoveTo(Home) so on.
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html
Low level functions: The closer it is to machine language, the lower-level the function is.
eg: Car.StartEngine(); or Car.ChangeGear();
using pandas.read_csv() is by, import pandas as pd -> pd.read_csv()

Tip2: pandas.read_csv() got its definition with many parameters and first parameter is 'required parameter' which we know because there is no default whereas the rest of these are 'optional parameters'. 

Then below usually will be short summary of the functions which will describe each one of the parameter which is super useful to read through and understand. 

So thats the simple example above of how we should use pandas document.

Question 1: example 2 
    Lets go back and do a different google search for "pandas drop". So when you are in that page 
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html in this page it says "pandas.Dataframe.drop" means drop() is a Dataframe method, in other words let say we have a Dataframe UFO and if we wanna use drop() method, then we gonna say "ufo.drop()" but not "pd.drop()" because drop() is a Dataframe method and not Pandas method.

We can contrast above example(Dataframe method) with a Series method. when we go to this page, we see that value_counts() is a Series method as "pandas.Series.value_count" https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html

Next thing we want to know about pandas document is " API Reference" which is most useful page in the pandas documentation.
API Reference page: https://pandas.pydata.org/pandas-docs/stable/reference/index.html
Basically, this page gives an overview of all public pandas objects, functions and methods.  

Would always encourage everyone to read " Whats New "https://pandas.pydata.org/pandas-docs/stable/whatsnew/index.html page either to catch up on the past couple versions or everytime a new version is released so we know what has changed. 

Other thing of note is : most of these are quite long tutorials essentially on topics like " Working with missing data" 
https://pandas.pydata.org/pandas-docs/stable/user_guide/missing_data.html  or "Group By" https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html  or "merge-join and concatenate" https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html  they tend to be really more to learn which has ton of information, but if we spent time to go through them we actually learn a lot about the topic and how to work with pandas.    

###### Question 2: Difference between ufo.isnull() and pd.isnull(ufo)?

When we search "pandas isnull" in Google, we will get 3 main searches(same version) one after the other. 

pandas.isnull :           https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.isnull.html

pandas.DataFrame.isnull : https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isnull.html

pandas.Series.isnull :    https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.isnull.html

pandas.isnull : Here isnull is called top-level function means its called by doing pd(which is pandas, pandas itself is higher level we are referring to).

Here isnull is top-level(pandas) method and Dataframe method and also Series method.

So there are different ways to use above 3 pandas isnull :: Lets see in code to understand clearly as below:

In [1]:
import pandas as pd

In [2]:
ufo = pd.read_csv('http://bit.ly/uforeports')

lets do two commands which are in above question:

In [3]:
ufo.head()

City Colors Reported Shape Reported State             Time
0                Ithaca             NaN       TRIANGLE    NY   6/1/1930 22:00
1           Willingboro             NaN          OTHER    NJ  6/30/1930 20:00
2               Holyoke             NaN           OVAL    CO  2/15/1931 14:00
3               Abilene             NaN           DISK    KS   6/1/1931 13:00
4  New York Worlds Fair             NaN          LIGHT    NY  4/18/1933 19:00

In [4]:
pd.isnull(ufo).head()

City  Colors Reported  Shape Reported  State   Time
0  False             True           False  False  False
1  False             True           False  False  False
2  False             True           False  False  False
3  False             True           False  False  False
4  False             True           False  False  False

Above command " pd.isnull(ufo) " : is checking for missing values and its True for some and False for others when we saw sample of data using head().

In [5]:
ufo.isnull().head()

City  Colors Reported  Shape Reported  State   Time
0  False             True           False  False  False
1  False             True           False  False  False
2  False             True           False  False  False
3  False             True           False  False  False
4  False             True           False  False  False

pd.isnull(ufo).head() and ufo.isnull().head() produced same result.So its little common in pandas to be able to access a function as a method of a Dataframe or a Series or as a top-level function. We can use it either whatever we prefer, either approach is fine. 

###### Question 3: Why are DataFrame slices inclusive when using .loc, but exclusive when using .iloc?

In [6]:
ufo.loc[0:4,:] # Pass it slice we want rows 0 to 4 and all columns

City Colors Reported Shape Reported State             Time
0                Ithaca             NaN       TRIANGLE    NY   6/1/1930 22:00
1           Willingboro             NaN          OTHER    NJ  6/30/1930 20:00
2               Holyoke             NaN           OVAL    CO  2/15/1931 14:00
3               Abilene             NaN           DISK    KS   6/1/1931 13:00
4  New York Worlds Fair             NaN          LIGHT    NY  4/18/1933 19:00

Above is a LABEL BASED INDEXING and label based indexing with loc[] is inclusive on both sides, so it includes both 0 and 4 and all columns. 

Whereas if we use iloc[] integer location mean POSITIONAL BASED INDEXING as below:

In [7]:
ufo.iloc[0:4,:]

City Colors Reported Shape Reported State             Time
0       Ithaca             NaN       TRIANGLE    NY   6/1/1930 22:00
1  Willingboro             NaN          OTHER    NJ  6/30/1930 20:00
2      Holyoke             NaN           OVAL    CO  2/15/1931 14:00
3      Abilene             NaN           DISK    KS   6/1/1931 13:00

Here we got 0, 1, 2, 3 rows because iloc[] with a slice is inclusive on the left and exclusive on the right. 

Lets see why DataFrame slices inclusive when using .loc, but exclusive when using .iloc  : :

   Pandas is built on top of Numpy and whenever possible pandas tries to follow Numpy conventions so iloc[] is exclusive on the right because that's how Numpy does its slicing. For instance : :

In [8]:
ufo.values # 'values' is a Dataframe attribute 

array([['Ithaca', nan, 'TRIANGLE', 'NY', '6/1/1930 22:00'],
       ['Willingboro', nan, 'OTHER', 'NJ', '6/30/1930 20:00'],
       ['Holyoke', nan, 'OVAL', 'CO', '2/15/1931 14:00'],
       ...,
       ['Eagle River', nan, nan, 'WI', '12/31/2000 23:45'],
       ['Eagle River', 'RED', 'LIGHT', 'WI', '12/31/2000 23:45'],
       ['Ybor', nan, 'OVAL', 'FL', '12/31/2000 23:59']], dtype=object)

Here 'values' which is a Dataframe attribute shows us underlying Numpy array of a Dataframe. And if we slice 'values'as below:

In [9]:
ufo.values[0:4,:]

array([['Ithaca', nan, 'TRIANGLE', 'NY', '6/1/1930 22:00'],
       ['Willingboro', nan, 'OTHER', 'NJ', '6/30/1930 20:00'],
       ['Holyoke', nan, 'OVAL', 'CO', '2/15/1931 14:00'],
       ['Abilene', nan, 'DISK', 'KS', '6/1/1931 13:00']], dtype=object)

Again here, we got rows 0, 1, 2, 3 and all columns. So in other words iloc[] is just copying what Numpy uses for slicing which is actually Numpy syntax and pandas reuses for iloc[], inclusive on the left and exclusive on the right. 

Now lets see why does NUMPY do inclusive of left and exclusive of right::

I think, that's how Python range does it. if we say as below:

In [10]:
list(range(0, 4))

[0, 1, 2, 3]

Here we will get 0, 1, 2, 3 because range inclusive on the left and exclusive on the right.

So PYTHON did it thus NUMPY did it thus PANDAS did it with iLoc[].

if iLoc[] works this way then why Loc[](inclusive of both sides) works bit different from iLoc[]. Let us see why that actually make sense here below:

In [11]:
ufo.head()

City Colors Reported Shape Reported State             Time
0                Ithaca             NaN       TRIANGLE    NY   6/1/1930 22:00
1           Willingboro             NaN          OTHER    NJ  6/30/1930 20:00
2               Holyoke             NaN           OVAL    CO  2/15/1931 14:00
3               Abilene             NaN           DISK    KS   6/1/1931 13:00
4  New York Worlds Fair             NaN          LIGHT    NY  4/18/1933 19:00

Lets say that we want to use Loc[] to select all rows and columns 'City' through 'State':

In [12]:
ufo.loc[:,'City':'State']

City Colors Reported Shape Reported State
0                    Ithaca             NaN       TRIANGLE    NY
1               Willingboro             NaN          OTHER    NJ
2                   Holyoke             NaN           OVAL    CO
3                   Abilene             NaN           DISK    KS
4      New York Worlds Fair             NaN          LIGHT    NY
...                     ...             ...            ...   ...
18236            Grant Park             NaN       TRIANGLE    IL
18237           Spirit Lake             NaN           DISK    IA
18238           Eagle River             NaN            NaN    WI
18239           Eagle River             RED          LIGHT    WI
18240                  Ybor             NaN           OVAL    FL

[18241 rows x 4 columns]

Above works just as expected and this tells us why slices are inclusive on both sides of Loc[ ]. Answer to this is,  if Loc[ ] was inclusive on the left and exclusive on the right, we would have to say " 'City':'Time' " to get 'City' through 'State' since it would exclude the last one('Time'). 

When we read the code  " 'City':'Time' " to get 'City' through 'State' , anyone would be puzzled because Loc[ ] is Label wise picking rows(or Columns). So whatever the rows/columns we require, we should only code for that using Labels of rows/columns.

When index is either .index(row labels of Dataframe) or .columns(column labels of Dataframe) are not consecutive numbers, its make more sense for Loc[ ] of these slices to be inclusive on both sides. 

To Summarize, the logic for Loc[ ] and iLoc[ ] both separately make sense for the reason they were invented. 
We should just remember the way both logic works independentley when we are doing slices.

###### Question 4: How do we randomly sample rows from large Dataframe / large csv files? ::

Lets see how we pick irregular rows or sampling, regardless of Dataframe size: 

In [13]:
ufo.sample(n=3)

City Colors Reported Shape Reported State             Time
10187  New York City             NaN          OTHER    NY   4/25/1997 0:00
8384        Ringgold             NaN       TRIANGLE    GA  7/16/1995 20:00
10599     New Jersey             NaN       FIREBALL    NJ  7/15/1997 21:00

Here we got 3 random rows and note that everytime we run the above sample() command, we will get different 3 random rows as below:

In [14]:
ufo.sample(n=3)

City Colors Reported Shape Reported State              Time
8769     New York City             RED            NaN    NY  10/11/1995 18:30
17818           Marion          ORANGE          LIGHT    KS  10/29/2000 21:30
18077  North Tonawanda             NaN          OTHER    NY   12/4/2000 21:40

In [15]:
ufo.sample(n=3)

City Colors Reported Shape Reported State             Time
5220  Forestville             NaN           DISK    CA   6/1/1985 17:30
1630     Pasadena             NaN          CIGAR    TX  6/15/1968 17:00
3154        Baker             NaN           OVAL    MT  2/20/1976 22:00

But if we are familiar with the concept of reproducibility, we can use random state parameter for reproducibility as below:

In [16]:
ufo.sample(n=3, random_state=64)

City Colors Reported Shape Reported State             Time
17860  Bloomfield             NaN           DISK    NJ  11/2/2000 19:45
7545    Lewisburg             NaN           DISK    TN  9/27/1994 23:20
10109      Juneau    ORANGE GREEN          LIGHT    AK  3/31/1997 21:00

In [17]:
ufo.sample(n=3, random_state=64)

City Colors Reported Shape Reported State             Time
17860  Bloomfield             NaN           DISK    NJ  11/2/2000 19:45
7545    Lewisburg             NaN           DISK    TN  9/27/1994 23:20
10109      Juneau    ORANGE GREEN          LIGHT    AK  3/31/1997 21:00

In [18]:
ufo.sample(n=3, random_state=64)

City Colors Reported Shape Reported State             Time
17860  Bloomfield             NaN           DISK    NJ  11/2/2000 19:45
7545    Lewisburg             NaN           DISK    TN  9/27/1994 23:20
10109      Juneau    ORANGE GREEN          LIGHT    AK  3/31/1997 21:00

The goal of setting a random_state is to ensure reproducibility, meaning that we will get the exact same results every time we run our code. However, it doesn't actually matter what number we use as the "seed" (that's the term use to refer to that number), as long as we don't change it in between runs.

Above we can see, when we use 'random_state' parameter, we will notice that every time we run it , we are getting same exact results. Now if we are running exact same line of code with " random_state = 64 ", we should be getting same results. So thats the REPRODUCIBILITY.

Lets say 'fraction of rows' instead of number of rows as below :

In [19]:
ufo.sample(frac=0.75, random_state=95) # 75% of the rows and random_state=95

City Colors Reported Shape Reported State             Time
7046    Gig Harbor             NaN           DISK    WA   6/1/1993 16:00
8153        Tacoma             NaN            NaN    WA   5/27/1995 0:54
15971     Portland             NaN            NaN    OR  2/27/2000 23:00
10269   Lake Tahoe             NaN      FORMATION    CA  5/24/1997 20:00
1716   Church Hill             NaN         SPHERE    PA  8/15/1968 15:00
...            ...             ...            ...   ...              ...
15424         Kent             NaN         CIRCLE    WA  12/2/1999 21:00
10655  Gulf Breeze          ORANGE      FORMATION    FL  7/30/1997 20:10
13612     Bellevue             NaN            NaN    WA    6/2/1999 7:50
13812     Florence             NaN          CIGAR    OR   7/1/1999 15:00
10036      Phoenix             NaN       TRIANGLE    AZ  3/13/1997 21:00

[13681 rows x 5 columns]

So thats the super useful that we can either specify number of rows or the fraction, however both can't be used at once.

###### Useful hint:  'Train-Test-Split' concept in Machine Learning

Train-Test-Split: Meaning we want certain % of rows in the Training set and remaining rows not overlapping in the Testing set, Lets see how to implement that in Pandas: 

In [20]:
train = ufo.sample(frac=0.75, random_state=95) # store random sample of 75% as train 

'ufo.index' is the index for 100% of the rows in the ufo DataFrame. 'train.index' is the index for 75% of the rows in the ufo DataFrame.

In [21]:
train.index

Int64Index([ 7046,  8153, 15971, 10269,  1716, 17622,  3671, 12323,  9592,
             9888,
            ...
             7584,  8593, 18075,  1379,  3485, 15424, 10655, 13612, 13812,
            10036],
           dtype='int64', length=13681)

Here the thing is we cannot just say 'frac=0.25' and get the other 25% of random sample which doesn't work that way. 

We require, non-overlapping(all the other rows which are not used in 'train' Dataframe) and random 25% as Testing set. Lets see how to do that here::

In [22]:
ufo.index.isin(train.index)

array([ True,  True,  True, ...,  True,  True,  True])

Above ufo.index.isin(train.index) outputs a Series of booleans (with the same length as the ufo DataFrame) that contains a True for every row in ufo that is also in train. Thus, it outputs a Series in which 75% of the elements are True (corresponding to which rows are in the train DataFrame).

In [23]:
test = ufo.loc[~ufo.index.isin(train.index),:]

The tilde(~) character inverts the Series, meaning Trues become Falses and Falses become Trues. At that point, the boolean Series can be passed to the loc method to select the other 25% of the rows from the ufo DataFrame.

In [24]:
test.shape

(4560, 5)

In [25]:
train.shape

(13681, 5)

In [26]:
ufo.shape

(18241, 5)

In [27]:
18241/100

182.41

In [28]:
182.41*75

13680.75

Here, did math to know the 75%data  from 18241 and got 13680.75(13681). so our 'train' row count is as expected.

In [29]:
182.41*25

4560.25

Here, to know the 25% data from 18241 and got 4560.25(4560). so our 'test' row count is also as expected.